sympy/ntheory/ec.py

```diff
@@ -32,28 +33,29 @@ class EllipticCurve():
 
     """
 
-    def __init__(self, a4, a6, a1=0, a2=0, a3=0, domain=QQ):
-        self._domain = domain
+    def __init__(self, a4, a6, a1=0, a2=0, a3=0, modulus = 0):
+        if modulus == 0:
+            self._domain = QQ
+        else:
+            self._domain = FF(modulus)
+        self.modulus = modulus
         # Calculate discriminant
         self._b2 = a1**2 + 4 * a2
         self._b4 = 2 * a4 + a1 * a3
         self._b6 = a3**2 + 4 * a6
         self._b8 = a1**2 * a6 + 4 * a2 * a6 - a1 * a3 * a4 + a2 * a3**2 - a4**2
         self._discrim = self._domain(-self._b2**2 * self._b8 - 8 * self._b4**3 - 27 * self._b6**2 + 9 * self._b2 * self._b4 * self._b6)
-        self._a1 = self._domain(a1)
-        self._a2 = self._domain(a2)
-        self._a3 = self._domain(a3)
-        self._a4 = self._domain(a4)
-        self._a6 = self._domain(a6)
+        self._a1 = sympify(self._domain(a1))
```

Comment:
These private coefficients should not be sympified. Working with coefficients belonging to the domain is more efficient.

If this is made a subclass of Basic, then the coefficients should be sympified and made public: `self.a1`, ... But there is no need to do that yet.

Reply:
FF(7)(5) returns `2 mod7`. But these are required in many operations and it gives error that this is not an integer. `as_int(FF(7)(5))` gives error.

Reply:
Why should you want to use `as_int` on elements of a finite field?

Reply:
It is used by other functions. Like  when calculating points, we call `quadratic congruence`, there it checks with `as_int`.  I suppose we could sympify  just before calling `quadratic congruence`

Reply:
>  we could sympify just before calling quadratic congruence

Yes, we can sympify the arguments.